### PR TITLE
[ISSUE-516][FEATURE] Worker should clean remaining directory when start before registering to Master

### PR DESCRIPTION
**Pull request by nafiyAix**
### What changes were proposed in this pull request?
Worker will clean remaining directory under working dir before registering Master.
image

### Why are the changes needed?
To avoid the issue that there isn't enough disk space and be put into blacklist.

### What are the items that need reviewer attention?
only test with local FS, and do we need to exit JVM or just print a warning log if the data can't be completely cleaned up in given time.

### Related issues.
#516 

### Related pull requests.
#515 

### How was this patch tested?

/cc @related-reviewer

/assign @main-reviewer

### CODE REVIEW

File: common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala [hunk 1]
```diff
@@ -778,6 +778,14 @@ object RssConf extends Logging {
     conf.getTimeAsSeconds("rss.worker.status.check.timeout", "30s")
   }
 
+  def checkIfFileCleanedRetryCount(conf: RssConf): Int = {
+    conf.getInt("rss.worker.checkIfFileCleanedRetryCount", 3)
```
Comment (AngersZhuuuu) [inline common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala:782]: `rss.worker.checkFileCleanRetryTimes`

Comment (AngersZhuuuu) [inline common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala:782]: `rss.worker.checkFileCleanRetryTimes`

File: common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala [hunk 2]
```diff
@@ -778,6 +778,14 @@ object RssConf extends Logging {
...
+  }
+
+  def checkIfFileCleanedTimeoutMs(conf: RssConf): Long = {
+    conf.getLong("rss.worker.checkIfFileCleanedTimeoutMs", 1000)
```
Comment (AngersZhuuuu) [inline common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala:786]: `rss.worker.checkFileCleanTimeoutMs`

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala [hunk 3]
```diff
@@ -170,15 +170,26 @@ final private[worker] class StorageManager(conf: RssConf, workerSource: Abstract
...
+      }
+    }
+    cleanupExpiredAppDirs(System.currentTimeMillis())
+    if(!checkIfWorkingDirCleaned) {
+      logWarning("Worker still has residual files in the working directory before registering with Master, " +
+        "please refer to the configuration document to increase rss.worker.checkIfFileCleanedRetryCount or " +
+        "rss.worker.checkIfFileCleanedTimeoutMs .")
+    } else {
+      logInfo("Successfully remove all files under working directory.")
     }
```
Comment (AngersZhuuuu) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala:193]: Only clean when not enable graceful shutdown?
```
 if (RssConf.workerGracefulShutdown(conf)) {
      try {
        val recoverFile = new File(RssConf.workerRecoverPath(conf), RECOVERY_FILE_INFOS_FILE_NAME)
        this.fileInfosDb = LevelDBProvider.initLevelDB(recoverFile, CURRENT_VERSION)
        reloadAndCleanFileInfos(this.fileInfosDb)
      } catch {
        case e: Exception =>
          logError("Init level DB failed:", e)
          this.fileInfosDb = null
      }
    } else {
      cleanupExpiredAppDirs(System.currentTimeMillis())
      if(!checkIfWorkingDirCleaned) {
        logWarning("Worker still has residual files in the working directory before registering with Master, " +
          "please refer to the configuration document to increase rss.worker.checkIfFileCleanedRetryCount or " +
          "rss.worker.checkIfFileCleanedTimeoutMs .")
      } else {
        logInfo("Successfully remove all files under working directory.")
      }
    }
```

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala [hunk 4]
```diff
@@ -437,6 +448,38 @@ final private[worker] class StorageManager(conf: RssConf, workerSource: Abstract
...
   }
 
+  private def checkIfWorkingDirCleaned: Boolean = {
+    var retryCount = 0
```
Comment (AngersZhuuuu) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala:452]: `retryTimes`

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala [hunk 5]
```diff
@@ -437,6 +448,38 @@ final private[worker] class StorageManager(conf: RssConf, workerSource: Abstract
...
+    var isLocalFsCleaned = true
+    var isHdfsCleaned = true
+    val awaitInterval = RssConf.checkIfFileCleanedTimeoutMs(conf)
```
Comment (AngersZhuuuu) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala:455]: Interval? Timeout?

**Review by AngersZhuuuu — COMMENTED** [review]
Pls run `./dev/reformat`

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala (hunk 3 above)
Comment (waitinfuture) [reply server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala:193]: We don't know if the last shutdown is a graceful shutdown, maybe this startup is the first time that graceful shutdown is enabled. Besides, even though last stop was graceful, there's a chance that data was not completed cleaned for reasons like timeout or others.

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala [hunk 6]
```diff
@@ -437,6 +448,38 @@ final private[worker] class StorageManager(conf: RssConf, workerSource: Abstract
     }
   }
 
+  private def checkIfWorkingDirCleaned: Boolean = {
```
[the remark this answers is not in the document]
Comment (nafiyAix) [reply server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala:451]: clean job has been done by     
```
cleanupExpiredAppDirs(System.currentTimeMillis())
```
this is checking if dir has been cleaned completely

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala [hunk 7]
```diff
@@ -437,6 +449,40 @@ final private[worker] class StorageManager(conf: RssConf, workerSource: Abstract
...
+    var hdfsCleaned = true
+    val awaitTimeout = RssConf.checkFileCleanTimeoutMs(conf)
+    while (retryTimes < RssConf.checkFileCleanRetryTimes(conf)) {
+      val isEmpty = !disksSnapshot().filter(_.status != DiskStatus.IoHang).exists { case diskInfo =>
+        diskInfo.dirs.exists { case workingDir =>
+          workingDir.listFiles().nonEmpty
+        }
+      }
+      localCleaned = isEmpty && localCleaned
+
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala:464]: We can define hdfsCleaned here.

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala [hunk 8]
```diff
@@ -437,6 +449,40 @@ final private[worker] class StorageManager(conf: RssConf, workerSource: Abstract
...
+          !hdfsFs.listFiles(new Path(hdfsDir, RssConf.workingDirName(conf)), false).hasNext &&
+            hdfsCleaned
+      }
+
+      if (localCleaned && hdfsCleaned) {
+        return true
+      }
+      retryTimes += 1
+      localCleaned = true
+      hdfsCleaned = true
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala:476]: We can remove this line then

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala [hunk 9]
```diff
@@ -170,15 +170,27 @@ final private[worker] class StorageManager(conf: RssConf, workerSource: Abstract
...
-    try {
-      val recoverFile = new File(RssConf.workerRecoverPath(conf), RECOVERY_FILE_INFOS_FILE_NAME)
-      this.fileInfosDb = LevelDBProvider.initLevelDB(recoverFile, CURRENT_VERSION)
-      reloadAndCleanFileInfos(this.fileInfosDb)
-    } catch {
-      case e: Exception =>
-        logError("Init level DB failed:", e)
-        this.fileInfosDb = null
+
+  {
```
[the remark this answers is not in the document]
Comment (nafiyAix) [reply server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala:174]: I think both of them are things that StorageManager has to do in the start phase, so put them together. How about put them into same one method like onStart/start?

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala [hunk 10]
```diff
@@ -170,15 +170,27 @@ final private[worker] class StorageManager(conf: RssConf, workerSource: Abstract
...
+        val recoverFile = new File(RssConf.workerRecoverPath(conf), RECOVERY_FILE_INFOS_FILE_NAME)
+        this.fileInfosDb = LevelDBProvider.initLevelDB(recoverFile, CURRENT_VERSION)
+        reloadAndCleanFileInfos(this.fileInfosDb)
+      } catch {
+        case e: Exception =>
+          logError("Init level DB failed:", e)
+          this.fileInfosDb = null
+      }
+    }
+    cleanupExpiredAppDirs(System.currentTimeMillis())
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala:186]: I just realized that we should avoid deleting shuffle data restored from leveldb, which is in fileInfos' keySet.

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (waitinfuture) [reply server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/storage/StorageManager.scala:174]: There are multiple pieces of initialization code in this class, many of them are inside the initialization body of fields, I think that way is fine.

**Comment by nafiyAix** [comment]
FYI: Add a check about the existence of workingDir, because when the worker is first initialized on localFs/hdfs , no existed workingDir. @waitinfuture

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

**Pull request merged**